piqture/neural_networks/qcnn.py

```diff
@@ -12,70 +12,93 @@
+                             neural network circuit. This determines the dimensions
+                             of the input that can be processed.
         """
         QuantumNeuralNetwork.__init__(self, num_qubits)
 
-    def sequence(self, operations: list[tuple[Callable, dict]]) -> QuantumCircuit:
+    def sequence(
+        self, operations: list[tuple[Type[BaseLayer], dict]]
+    ) -> QuantumCircuit:
         """
-        Builds a QNN circuit by composing the circuit with given
-        sequence of list of operations.
+        Build a QNN circuit by composing the circuit with given sequence of operations.
 
         Args:
-            operations (list[tuple[Callable, dict]]: a tuple
-            of a Layer object and a dictionary of its arguments.
+            operations (list[tuple[Type[BaseLayer], dict]]): A list of tuples where
+                      each tuple contains a Layer class that inherits from BaseLayer
+                      and a dictionary of its arguments.
 
         Returns:
-            circuit (QuantumCircuit): final QNN circuit with all the
-            layers.
+            QuantumCircuit: Final QNN circuit with all the layers applied.
+
+        Raises:
+            TypeError: If operations format is invalid or if any operation doesn't
+                      inherit from BaseLayer.
+            ValueError: If operations list is empty.
         """
+        # Validate operations list
         if not isinstance(operations, list):
             raise TypeError("The input operations must be of the type list.")
 
+        if not operations:
+            raise ValueError("The operations list cannot be empty.")
+
         if not all(isinstance(operation, tuple) for operation in operations):
             raise TypeError(
                 "The input operations list must contain tuple[operation, params]."
             )
 
-        if not callable(operations[0][0]):
-            raise TypeError(
-                "Operation in input operations list must be Callable functions/classes."
-            )
-
-        if not isinstance(operations[0][1], dict):
-            raise TypeError(
-                "Parameters of operation in input operations list must be in a dictionary."
-            )
-
+        # Validate each operation
+        for idx, (layer, params) in enumerate(operations):
+            # Check if it's a class and inherits from BaseLayer
+            if not isclass(layer):
+                raise TypeError(
+                    f"Operation at index {idx} must be a class, got {type(layer).__name__}"
+                )
+
+            if not issubclass(layer, BaseLayer):
+                raise TypeError(
+                    f"Operation at index {idx} must inherit from BaseLayer, got {layer.__name__}"
+                )
+
+            # Prevent using BaseLayer itself
+            if layer is BaseLayer:
+                raise TypeError(
+                    f"Operation at index {idx} cannot be BaseLayer itself, got *"
```

Comment:
What does the asterisk denote here? We can simply have:
```suggestion
                    f"Operation at index {idx} cannot be BaseLayer itself."
```